First version. Record of design decisions for Paas billing rewrite (#126)
alphagov/paas-billing: docs/architecture/ADR002.md

> # ADR002: Decouple what we're calculating bills for from how the bills are calculated

Reviewer: Is this what it means, or is it more than whether or not they're already a tenant?
```suggestion
# ADR002: Decouple how the bills are calculated from whose bills they are
```

## Status

Accepted.

## Context

Tenants provision a wide variety of services and AWS/Aiven resources via Paas. We need to calculate bills for these services/resources. Paas billing receives records of when services/resources are created/renamed/deleted from upstream in the form of events from Cloudfoundry.

Using the original Paas billing system, bills are calculated based on a complex series of processing steps arriving at a set of bills. Bills are compiled individually for each tenant and for each month and these bills are stored in the database. The job to calculate these bills runs once a month.

In the Paas billing rewrite, this has been changed. We want to calculate bills for variable time periods and also to forecast bills for the future (for the billing calculator). However, the method to calculate the actual bill always needs to be the same.

## Decision

> The code to calculate the bill has been decoupled from the code used to calculate the bill. This is so we can use exactly the same code for calculation of all bills, whether these bills are for tenants or for prospective tenants (using the billing calculator).

Reviewer: ```suggestion
The calculation of bills needs to be decoupled from the bill owner. This is so we can use exactly the same code for calculation of all bills, whether these bills are for tenants or for prospective tenants (using the billing calculator).
```

The approach taken is:
    1. Populate database temporary table with what is being billed (which resources over which time interval, including the future). This can be in a stored function or embedded SQL. This is the code entry point into billing.

>     2. Call a stored function to calculate the bill ([`calculate_bill`](../../billing-db/sprocs/calculate_bill.sql)) only using the contents of the temporary table populated in step 1 above. No parameters are passed into ([`calculate_bill`](../../billing-db/sprocs/calculate_bill.sql)).

Reviewer: I was going to say that this link wouldn't work in the Team Manual. But I've now realised that this is not the team manual repo! Should these ADRs be [in the paas-team-manual like the others](https://github.com/alphagov/paas-team-manual/tree/main/source/architecture_decision_records)?

Author: No, this follows the convention in some other repos where we have ADRs in the code repo so the reasons why certain design decisions were made are recorded

## Consequences

> Importantly, this means that we can have a single source of truth for calculating all Paas bills. Tenants' bills or prospective tenants' bills, the latter using the billing calculator, will be calculated using the same code.

Reviewer: ```suggestion
There is a single source of truth for calculating all GOV.UK PaaS bills. Tenants' bills or prospective tenants' bills, the latter using the billing calculator, will be calculated using the same code.
```
